### ISSUE TITLE
Change the output of the predict functions

### ISSUE THREAD

**Issue by istvanp1** [skpref/skpref#50]
- Make predict_task() return an appropriate PosetVector type.
- Make predict_proba_task() return a distribution object.
    - Create a distribution object, this will store the function for calculating the probability of something happening, but will not calculate all permutation, which would take a long time.
    - Write a method for it called pmf, which takes an outcome of interest as the input and returns a probability for it `python pmf(outcome: PosetVector) -> float`

So now in a pairwise comparison, the journey would look like the following:

```python
# after appropriate functions have been imported.
>>> mybt.fit_task(NBA_results_task_train)
>>> my_predictions = mybt.predict_task(NBA_results_task_predict)
>>> my_predictions.top
np.arrray(['A', 'B', ...])
>>> my_predictions.boot
np.arrray(['X', 'A', ...])
>>> my_prob_dist_for_predictions = mybt.predict_proba_task(NBA_results_task_predict)
>>> my_prob_dist_for_predictions.pmf(outcome='A', obs=1')
0.8
# This will give 0 to cases where 'A' was not a choice.
>>> my_prob_dist_for_predictions.pmf(outcome='A')
np.array([0.8, 0, 0, 0.1...])
>>> my_prob_dist_for_predictions.pmf(column='team1')
np.array([0.8, 0.4, 0.7...])
>>> my_prob_dist_for_predictions.pmf(column='team2')
np.array([0.2, 0.6, 0.3...])
```

More details in [this wiki page](https://github.com/skpref/skpref/wiki/What-should-be-the-output-of-predict-functions%3F).

Steps (estimated completion date November 29):
- [x] Make the classification reducer return a SubsetPosetVector for predict_task when aggregating to pairwise comparisons
- [x] Make the classification reducer return a SubsetPosetVector for predict_task when aggregating to a subset choice
- [x] Make Bradley terry return a SubsetPosetVector for predict_task
- [x] Write predict_proba_task for the classifier reducer for a pairwise comparison
- [x] Write predict_proba_task for Bradley Terry for a pairwise comparison
- [ ] Write predict_proba_task for the classifier reducer for a discrete choice
- [ ] Write predict_proba_task for Bradley Terry for a discrete choice
- [ ] Write a unit test for predict_proba_task()

related [branch](https://github.com/skpref/skpref/tree/50-change-the-output-of-the-predict-functions)

**Comment by mloning** [comment]
What's your idea for the distribution object?

**Comment by istvanp1** [comment]
They will be discrete distribution objects similar to scipy.stats.rv_discrete, the original example was incorrect, I have now edited it so that my_prob_dist_for_predictions has a method called pmf() which takes outcome, column and obs parameters.

**Comment by istvanp1** [comment]
When I mean distribution I don't mean a specifically known distribution like Poisson. I mean the distribution of the probabilities for each result. Similar to what comes out of predict_proba() in sklearn.

**Comment by istvanp1** [comment]
I don't think anymore that we should have an object for distributions. Currently, this object would only have one method which is pmf() to return the probabilities. We should just make predict_proba_task() act like pmf.

**Comment by istvanp1** [comment]
We would have the following:

```python
# after appropriate functions have been imported.
>>> mybt.fit_task(NBA_results_task_train)
>>> my_predictions = mybt.predict_task(NBA_results_task_predict)
>>> my_predictions.top
np.arrray(['A', 'B', ...])
>>> my_predictions.boot
np.arrray(['X', 'A', ...])
>>> mybt.predict_proba_task(NBA_results_task_predict, outcome='A, obs=1')
0.8
# This will give 0 to cases where 'A' was not a choice.
>>> mybt.predict_proba_task(NBA_results_task_predict, outcome='A)
np.array([0.8, 0, 0, 0.1...])
>>> mybt.predict_proba_task(NBA_results_task_predict, column='team1')
np.array([0.8, 0.4, 0.7...])
>>> mybt.predict_proba_task(NBA_results_task_predict, column='team2')
np.array([0.2, 0.6, 0.3...])
```

**Comment by istvanp1** [comment]
And potentially the output types would make more sense as dictionaries.
```python
# after appropriate functions have been imported.
>>> mybt.fit_task(NBA_results_task_train)
>>> my_predictions = mybt.predict_task(NBA_results_task_predict)
>>> my_predictions.top
np.arrray(['A', 'B', ...])
>>> my_predictions.boot
np.arrray(['X', 'A', ...])
>>> mybt.predict_proba_task(NBA_results_task_predict, outcome='A', obs=1')
{'obs': [1], 
 'A': np.array([0.8])}
# This will give 0 to cases where 'A' was not a choice.
>>> mybt.predict_proba_task(NBA_results_task_predict, outcome='A')
{'A': np.array([0.8, 0.01, 0, 0.1...])}
>>> mybt.predict_proba_task(NBA_results_task_predict, column='team1')
{'team1 chosen': np.array([0.8, 0.99, 0.7...])}
>>> mybt.predict_proba_task(NBA_results_task_predict, column='team2')
{'team2 chosen': np.array([0.2, 0.01, 0.3...])}
>>> mybt.predict_proba_task(NBA_results_task_predict, outcome=['A', 'B'])
{'A': np.array([0.8, 0.01, 0, 0.1...]),
 'B': np.array([0.2, 0.99, 0, 0.5...]}
```

**Comment by fkiraly** [comment]
The main problem with storing the pmf explicitly by vectors of probabilities is what we discussed a while ago: the number of events the distribution is a distribution over may be huge and combinatorially explosive - think of distributions over sub-sets of a 100-element set (of products etc), there are about a quintillion of these. That's about the number of bytes of data created in the world daily by common estimates, and requires about a trillion normal sized hard drives to store.

It's of course feasible if you have a binary choice, but if we also think about the interface covering sub-set choice and ranking, you are inviting a problem down the line...

Whereas, as long as the distribution over a huge domain is simple enough to represent - e.g., as independent probabilities, or a distribution of low combinatorial degree - the problem does not occur. But, for representation, I don't see a way to get around objects for distributions if predictions are to be stored in isolation.

Opinions, @mloning?

**Comment by fkiraly** [comment]
PS: you also say that the only method you want to call of the distribution is `pmf` - I do not see that to be the case, you at least also may want to retrieve:
- a symbolic description of the form, and/or parameters of it
- analytic methods for terms occurring in common probabilistic prediction losses, e.g., the "sum-of-squared-probabilities" for the Brier loss

**Comment by istvanp1** [comment]
predict_proba_task does not store all of the potential permutations but calculates the specific permutation requested by the user. So the only way a quintillion permutations are predicted is if the user enters a quintillion different permutations in predict proba.

**Comment by istvanp1** [comment]
Regarding pmf, the parameters that generate the probabilities such as the latent strength parameters for the teams or the parameters on the covariates can be queried in another function.
